src/main/java/uk/gov/ch/developer/docs/controller/developer/SignInController.java

```diff
@@ -0,0 +1,231 @@
+
+        return email;
+    }
+
+    /**
+     * Constructs the authorisation URI with additional force and hint
+     * 
+     * @param originalRequestUri Original URI from which to redirect
+     * @param scope Scope of the request
+     * @param nonce
+     * @param email Email address from the session, empty if not present
+     * @return Authorisation URI
+     */
+    protected String createAuthoriseURIWithForceAndHint(String originalRequestUri, String scope,
+            String nonce, String email) {
+
+        StringBuilder sb = new StringBuilder(createAuthoriseURI(originalRequestUri, scope, nonce));
+        sb.append("&reauthenticate=force");
+        sb.append("&hint=");
+        sb.append(jweEncodeWithNonce(email, nonce, "email"));
+
+        return sb.toString();
+    }
+
+    /**
+     * Constructs the authorisation URI
+     * 
+     * @param originalRequestUri Original URI from which to redirect
+     * @param scope Scope of the request
+     * @param nonce
+     * @return Authorisation URI
+     */
+    protected String createAuthoriseURI(String originalRequestUri, String scope, String nonce) {
+
+        StringBuilder sb = new StringBuilder();
+        sb.append(authorizationUri);
+        sb.append("?");
+        sb.append("client_id=");
+        sb.append(clientId);
+        sb.append("&redirect_uri=");
+        sb.append(redirectUri);
+        sb.append("&response_type=code");
+
+        if (scope != null) {
+            sb.append("&scope=");
+            sb.append(scope);
+        }
+
+        sb.append("&state=");
+        sb.append(jweEncodeWithNonce(originalRequestUri, nonce, "content"));
+
+        return sb.toString();
+    }
+
+    /**
+     * Generates a secure unique key
+     * 
+     * @return Base64 encoded unique key
+     */
+    private String generateNonce() {
+
+        SecureRandom random = new SecureRandom();
+        byte[] bytes = new byte[5];
+        random.nextBytes(bytes);
+
+        return Base64.encodeBase64URLSafeString(bytes);
+    }
+
+    /**
+     * Encodes a URI with a nonce according to a JWE encoding algorithm
+     * 
+     * @param returnUri
+     * @param nonce
+     * @param attributeName
+     * @return JWE encoded string, comprised of the return URI and a nonce
+     */
+    protected String jweEncodeWithNonce(String returnUri, String nonce, String attributeName) {
+
+        JSONObject payloadJson = new JSONObject();
+        payloadJson.put(attributeName, returnUri);
+        payloadJson.put("nonce", nonce);
+
+        Payload payload = new Payload(payloadJson);
+        JWEHeader header = new JWEHeader(JWEAlgorithm.DIR, EncryptionMethod.A128CBC_HS256);
+        JWEObject jweObject = new JWEObject(header, payload);
+
+        byte[] key = Base64.decodeBase64(base64Key);
+
+        try {
+            jweObject.encrypt(new DirectEncrypter(key));
+        } catch (JOSEException e) {
+            LOGGER.error(e, null);
+        }
+
+        return jweObject.serialize();
```

Comment:
Is it risky to return the object if there was an exception thrown one the encryption?

Reply:
During the encryption or during the serialisation? As during the encryption an exception would be caught and logged out.

Reply:
During the encryption, the exception is caught and logged, but the code then continues to serialise the object for return. I would be much happier if, in the `catch` the code returned an empty result, or an error result, rather than dropping through to the `return jweObject.serialize` in case of potential exposure of elements that should be encrypted.

Reply:
I have returned an empty string after logging the exception.

Reply:
Oh no! David hates multiple exit points.